pkg/controllers/jobset_controller.go

```diff
@@ -323,37 +323,76 @@ func (r *JobSetReconciler) suspendJobSet(ctx context.Context, js *jobset.JobSet,
 	})
 }
 
-func (r *JobSetReconciler) resumeJobSetIfNecessary(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs) error {
+func (r *JobSetReconciler) resumeJobSetIfNecessary(ctx context.Context, js *jobset.JobSet, ownedJobs *childJobs, replicatedJobStatus []jobset.ReplicatedJobStatus) error {
 	log := ctrl.LoggerFrom(ctx)
 
 	nodeAffinities := map[string]map[string]string{}
+	replicatedJobToActiveJobs := map[string][]*batchv1.Job{}
 	for _, replicatedJob := range js.Spec.ReplicatedJobs {
 		nodeAffinities[replicatedJob.Name] = replicatedJob.Template.Spec.Template.Spec.NodeSelector
 	}
 
+	for _, job := range ownedJobs.active {
+		replicatedJobName := job.Labels[jobset.ReplicatedJobNameKey]
+		replicatedJobToActiveJobs[replicatedJobName] = append(replicatedJobToActiveJobs[replicatedJobName], job)
+	}
+
+	getReplicatedJobStatus := func(replicatedJobStatus []jobset.ReplicatedJobStatus, replicatedJobName string) jobset.ReplicatedJobStatus {
+		for _, rjobStatus := range replicatedJobStatus {
+			if replicatedJobName == rjobStatus.Name {
+				return rjobStatus
+			}
+		}
+		return jobset.ReplicatedJobStatus{}
+	}
+
 	// If JobSpec is unsuspended, ensure all active child Jobs are also
 	// unsuspended and update the suspend condition to true.
-	for _, job := range ownedJobs.active {
-		if ptr.Deref(job.Spec.Suspend, false) {
-			if job.Status.StartTime != nil {
-				job.Status.StartTime = nil
-				if err := r.Status().Update(ctx, job); err != nil {
-					return err
-				}
+	for idx, replicatedJob := range js.Spec.ReplicatedJobs {
+		jobStatus := getReplicatedJobStatus(replicatedJobStatus, replicatedJob.Name)
+		// Startup Policy for suspended will resume replicated jobs in order
+		// And then it will only resume the next set of jobs
+		// if the previous jobs are ready.
+		if inOrderStartupPolicy(js.Spec.StartupPolicy) {
+			isPreviousReplicatedJobReady := true
+			if idx > 0 {
+				prevJobStatus := getReplicatedJobStatus(replicatedJobStatus, js.Spec.ReplicatedJobs[idx-1].Name)
+				isPreviousReplicatedJobReady = replicatedJobsAtLeastReady(js.Spec.ReplicatedJobs[idx-1].Replicas, prevJobStatus)
 			}
-			if job.Labels != nil && job.Labels[jobset.ReplicatedJobNameKey] != "" {
-				// When resuming a job, its nodeSelectors should match that of the replicatedJob template
-				// that it was created from, which may have been updated while it was suspended.
-				job.Spec.Template.Spec.NodeSelector = nodeAffinities[job.Labels[jobset.ReplicatedJobNameKey]]
-			} else {
-				log.Error(nil, "job missing ReplicatedJobName label")
+			if !isPreviousReplicatedJobReady && !replicatedJobsAtLeastReady(replicatedJob.Replicas, jobStatus) {
+				return r.ensureCondition(ctx, js, corev1.EventTypeNormal, generateStartupPolicyCondition(false, replicatedJob.Name))
 			}
-			job.Spec.Suspend = ptr.To(false)
-			if err := r.Update(ctx, job); err != nil {
-				return err
+		}
```

Comment:
I can't really figure out a way to tell difference between created and ready with status.

I ended up checking the previous job to determine if I should create. And on the 0th element I create.

This seems to work e2e and currently I am working on the tests.

Reply:
When I try to not check the previous job, I usually end up having none of the jobs create.

Reply:
I am still not following. Can you tell me in what case the following logic fails to unsuspend the jobs in order? 

```
for idx, replicatedJob := range js.Spec.ReplicatedJobs {
    rjStatus = replicatedJobStatus[idx]
    If rjStarted && inOrderStartupPolicy(js.Spec.StartupPolicy) {
      continue;
    }

    for _, job := range replicatedJobToActiveJobs[replicatedJob.Name] {
	if err := r.resumeJob(ctx, job, nodeAffinities); err != nil {
		return err
	}
    }

    if inOrderStartupPolicy(js.Spec.StartupPolicy) {
      return r.ensureCondition(ctx, js, corev1.EventTypeNormal, generateStartupPolicyCondition(false, replicatedJob.Name))
    }
}
```

Reply:
can you please respond to the comment above: https://github.com/kubernetes-sigs/jobset/pull/246#discussion_r1472282417

The logic above is very simpler and I believe it should work, I didn't get a clear answer why you are not considering.

Reply:
Yea I'll update that and see if I can that to work.

Reply:
So the main issue you run into is you would never resume in that case. Since a suspended job is never ready.

That code would just keep continueing because you never resume the job.

Reply:
> Since a suspended job is never read

And that is correct, we don't want a different behavior.

if rjStatus is false, then we would not be continuing, we should be going through the resume loop of that replicatedJob and then exit the top level loop.

Reply:
This is what I get for trying to write code on my laptop while in an airport..

I’ll play around with this hopefully this week but def next! I am actually traveling to FOSDEM to talk about JobSet and Job work so I was feeling inspired while I was working on my presentation.

thank you for your reviews so far!

Reply:
Okay. You were correct! Once I had the integration tests working, it was easier to deduce any bugs I had. Pushed up a change for both suspend and creation without relying on previous status.